File: contrib/src/dhtnet/guard_peer_response.pl
#!/usr/bin/env perl

use strict;
use warnings;

my $guard = "    if (!response.from)\n        return;\n";

sub add_guard {
    my ($source) = @_;
    return ($source, 0) if index($source, '!response.from') >= 0;

    my $signature = 'ConnectionManager::Impl::onPeerResponse';
    my $signature_index = index($source, $signature);
    die "$signature definition not found\n" if $signature_index < 0;

    my $body_index = index($source, '{', $signature_index);
    die "$signature body not found\n" if $body_index < 0;

    my $body_prefix = substr($source, $signature_index, $body_index - $signature_index);
    die "$signature does not use the expected response parameter\n"
        unless $body_prefix =~ /\bresponse\b/;

    my $newline = index($source, "\r\n") >= 0 ? "\r\n" : "\n";
    my $insert = $newline . $guard;
    $insert =~ s/\n/$newline/g if $newline ne "\n";

    substr($source, $body_index + 1, 0) = $insert;
    return ($source, 1);
}

sub self_test {
    my $sample = <<'EOF';
void
ConnectionManager::Impl::onPeerResponse(PeerConnectionRequest&& response)
{
    auto deviceId = response.from->getId();
}
EOF

    my ($patched, $changed) = add_guard($sample);
    die "guard was not inserted\n" unless $changed;
    die "guard missing\n" unless index($patched, "if (!response.from)\n        return;\n") >= 0;
    die "guard inserted after dereference\n"
        unless index($patched, 'if (!response.from)') < index($patched, 'response.from->getId()');

    my ($second, $changed_again) = add_guard($patched);
    die "guard insertion is not idempotent\n" if $changed_again;
    die "idempotent patch changed source\n" unless $second eq $patched;
}

sub patch_file {
    my ($path) = @_;
    open(my $in, '<:encoding(UTF-8)', $path) or die "cannot read $path: $!\n";
    local $/;
    my $source = <$in>;
    close($in);

    my ($patched, $changed) = add_guard($source);
    if ($changed) {
        open(my $out, '>:encoding(UTF-8)', $path) or die "cannot write $path: $!\n";
        print {$out} $patched;
        close($out);
    }
}

if (@ARGV == 1 && $ARGV[0] eq '--self-test') {
    self_test();
    exit 0;
}

die "usage: $0 <connectionmanager.cpp> | --self-test\n" unless @ARGV == 1;
patch_file($ARGV[0]);

File: contrib/src/dhtnet/guard_peer_response.py
#!/usr/bin/env python3

import argparse
import re
import sys
from pathlib import Path


GUARD = """\
    if (!response.from)
        return;
"""


def add_guard(source):
    if "!response.from" in source:
        return source, False

    signature = "ConnectionManager::Impl::onPeerResponse"
    signature_index = source.find(signature)
    if signature_index < 0:
        raise RuntimeError(f"{signature} definition not found")

    body_index = source.find("{", signature_index)
    if body_index < 0:
        raise RuntimeError(f"{signature} body not found")

    body_prefix = source[signature_index:body_index]
    if not re.search(r"\bresponse\b", body_prefix):
        raise RuntimeError(f"{signature} does not use the expected response parameter")

    insert_index = body_index + 1
    newline = "\r\n" if "\r\n" in source else "\n"
    guard = newline + GUARD.replace("\n", newline)
    return source[:insert_index] + guard + source[insert_index:], True


def patch_file(path):
    source_path = Path(path)
    with source_path.open("r", encoding="utf-8", newline="") as source_file:
        source = source_file.read()
    patched, changed = add_guard(source)
    if changed:
        with source_path.open("w", encoding="utf-8", newline="") as source_file:
            source_file.write(patched)
    return changed


def self_test():
    sample = """\
void
ConnectionManager::Impl::onPeerResponse(PeerConnectionRequest&& response)
{
    auto deviceId = response.from->getId();
}
"""
    patched, changed = add_guard(sample)
    assert changed
    assert "if (!response.from)\n        return;\n" in patched
    assert "if (!response.from)" in patched.split("response.from->getId()")[0]

    second, changed_again = add_guard(patched)
    assert not changed_again
    assert second == patched


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("source", nargs="?")
    parser.add_argument("--self-test", action="store_true")
    args = parser.parse_args()

    if args.self_test:
        self_test()
        return 0

    if not args.source:
        parser.error("source path is required unless --self-test is used")

    try:
        patch_file(args.source)
    except Exception as exc:
        print(f"guard_peer_response.py: {exc}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
